agents-api/agents_api/common/utils/task_validation.py

```diff
@@ -0,0 +1,531 @@
+        or (expr.startswith("_.") and expr.endswith("]"))
+        or "outputs[" in expr
+        or "inputs[" in expr
+        or expr == "_"
+    ):
+        return "$ " + expr  # Keep space after $ for backward compatibility
+
+    return expr
+
+
+def validate_py_expression(
+    expr: str | None,
+    allow_placeholder_variables: bool = True,
+    expected_variables: set[str] | None = None,
+) -> dict[str, list[str]]:
+    """
+    Statically validate a Python expression before task execution.
+
+    Args:
+        expr: The Python expression to validate (or None)
+        allow_placeholder_variables: Whether to allow _ and placeholders like _.attr in expression
+        expected_variables: Optional set of expected variable names that should be available
+
+    Returns:
+        Dict with potential issues categorized by type
+    """
+    issues: dict[str, list[str]] = {
+        "syntax_errors": [],
+        "undefined_names": [],
+        "unsafe_operations": [],
+        "potential_runtime_errors": [],
+    }
+
+    # Skip None or empty expressions
+    if expr is None or not expr or not expr.strip():
+        return issues
+
+    # Apply backwards compatibility transformation first
+    expr = backwards_compatibility(expr)
+
+    # Ensure the expression is stripped before checking prefix
+    expr = expr.strip()
+
+    # Handle expressions with $ prefix - return early if it doesn't start with $
+    if not expr.startswith("$"):
+        return issues
+
+    # The space after $ is required for expressions to be evaluated as Python code
+    if not expr.startswith("$ "):
+        return issues
+
+    # Remove $ and strip any leading space after $
+    expr = expr[1:].strip()
+
+    # Special case: just a $ sign with nothing after it
+    if not expr:
+        return issues
+
+    # Handle f-string expressions (these are often used in templates)
+    expr = expr.strip()
+    if expr.startswith(("f'''", 'f"""')):
+        # Just basic syntax check for f-strings, can't do much static analysis
+        try:
+            ast.parse(expr)
+        except SyntaxError as e:
+            issues["syntax_errors"].append(f"F-string syntax error: {e!s}")
+        return issues
+
+    # Try to parse the expression to check for syntax errors
+    try:
+        tree = ast.parse(expr, mode="eval")
+    except SyntaxError as e:
+        issues["syntax_errors"].append(f"Syntax error: {e!s}")
+        return issues  # Return early if we can't even parse the expression
+
+    # Get all name references in the expression
+    name_nodes = [node for node in ast.walk(tree) if isinstance(node, ast.Name)]
+
+    # Check for undefined names
+    referenced_names = {
+        node.id
+        for node in name_nodes
+        if isinstance(node, ast.Name) and isinstance(node.ctx, ast.Load)
+    }
+
+    # Import ALLOWED_FUNCTIONS and stdlib from where they're defined
+    from ...activities.utils import ALLOWED_FUNCTIONS, stdlib
```

Comment:
can this be imported at the beginning of the file?

Reply:
causes a circular import :-/